[Hunting] Re-factor Hunting Library Code

terrancedejesus opened the pull request:
<!--
Thank you for your interest in and contributing to Detection Rules!
There are a few simple things to check before submitting your pull request
that can help with the review process. You should delete these items
from your submission, but they are here to help bring them to your attention.
-->
# Pull Request

*Issue link(s)*:
* https://github.com/elastic/ia-trade-team/issues/280

<!--
  Add Related Issues / PRs for context. Eg:
    Related to elastic/repo#999
    Resolves #123
  If there is no issue link, take extra care to write a clear summary and label the PR just as you would label an issue to give additional context to reviewers.
-->

## Summary - What I changed
Re-factors threat hunting library code for modularity and expanded functionality for hunters. Below is an incomplete list of changes:

* Add an `index.yml` to capture and store available hunts, UUIDs, etc.
* change hunting to be it's own package with CLI
* update `generate-markdown` functionality to allow single file path, folder (i.e. `aws`)
* add `refresh index` command to update the index
* add a `search` command to allow MITRE (+integration)-based searching for hunters
* added a `run-query` command to run a query against elasticsearch search API and let users know if matches exist or not
* added a `view-hunt` command to show contents of a hunting file in JSON or TOML format
* updated README

<details>
<summary>

## Additional Information

</summary>

### Fixing Okta folder structure
From https://github.com/elastic/detection-rules/pull/4064, the folder structure was merged in as `okta/docs/docs` and `okta/docs/queries`. As a result, when we would create the index files, `docs` was the header and it also broke searching. To fix this, we adjusted the filer structure to be `okta/docs` and `okta/queries` as expected.

### Adding `index.yml`
Prior to this PR, we had `index.md` which was a markdown file that stored an index of all of our hunts. This allowed visitors to see a centralized list of available hunting queries grouped by data source. However, for programmatic use in logic, it works better to also have an data-based index that can be easily referenced and loaded, thus the introduction of `index.yml`.

In this index, we have the following...

- UUID
- Hunting Analytic Name
- Location (File Path)
- MITRE (Technique & Sub-Technique IDs)

This YAML file is then used for refreshing the index and the markdown file, identifying file locations, also making it easier to search for specific queries based on MITRE information, such as the new `search` command. This will be the source of truth for hunting queries library.

We have also added `refresh-index` command which updates the YAML index from all hunting queries and then updates the markdown. 

<img width="935" alt="Screenshot 2024-09-18 at 2 52 21 PM" src="https://github.com/user-attachments/assets/b2bedf3d-a986-44e3-a69b-5b484bb9306d">

### Add CLI to Hunting Module
Prior to this PR, `hunting/` was a top-level module. An additional `__main__.py` and `click` module allows us to call the module from the top-level of the repo as such: `python -m hunting`. As such it is similar to `rta` and stands as a new feature of the detection rules repository. However, this still allows us to leverage modules or libraries from within `detection_rules` as we do with `attack.py` to not have so much redundancy. Commands can be added to `__main__.py` if we want to extend features and capabilities.

There are other OOP changes such as adding `utils.py` and `definitions.py` to replicate the same structure as the `detection_rules` module but keep it simple.

<img width="1202" alt="Screenshot 2024-09-18 at 2 51 00 PM" src="https://github.com/user-attachments/assets/3435d841-8b91-4955-a35c-5c95396d3f47">

### Adjust Markdown Generation
Prior to this PR, we were only capable of generating the markdown from TOML files for ALL queries at once, which required us to load all TOML hunting files and overwrite markdown files which was unnecessary. Rather, we have added `generate-markdown` command which then calls `process_toml_files` with parameters for both a single TOML filepath, a data source string such as `aws` or simply all docs. 

<img width="1238" alt="Screenshot 2024-09-18 at 2 50 32 PM" src="https://github.com/user-attachments/assets/ca62d2a1-10e1-4bde-b8ca-b1d4b205e194">

### Added a MITRE `search` command
While the repository is meant to be a simplistic representation of all our hunting queries, allowing visitors to copy and paste from markdown or programmatically load TOML files, as we build out more queries and docs, searching for the right hunt may be daunting. As search a simple and effective approach to enable visitors is to allow them to search for specific hunting queries based on the following:

- Tactic ID -  Returns all hunting queries whose techniques or sub-techniques fall within the bucket of that tactic ID
- Technique ID - Returns all hunting queries whose technique IDs match or sub-techniques whom fall into the bucket of that technique
- Subtechnique ID - Returns all hunting queries whose sub-technique IDs match
- Data Source - Allow an additional filter for data source, such as `aws` or `okta`. This proves useful if you want to search, for example, all TA0001 (Credential Access) related to Okta.

<img width="1956" alt="Screenshot 2024-09-18 at 2 48 41 PM" src="https://github.com/user-attachments/assets/aa4c8b19-214f-47e5-88a1-c76afc94dc45">

### Run Query
In order to do this, you must add a `.detection-rules-cfg.yaml` config with your Cloud ID and API key for authentication.

`python -m hunting run-query --file-path /Users/tdejesus/code/src/detection-rules/hunting/linux/queries/low_volume_external_network_connections_from_process.toml`

<img width="1259" alt="Screenshot 2024-09-25 at 2 28 49 PM" src="https://github.com/user-attachments/assets/ed7cc1cc-a0e2-43a1-b6f3-7716f6ce032a">

`python -m hunting run-query --uuid ef579900-75ef-11ef-b47f-f661ea17fbcc `
<img width="1013" alt="Screenshot 2024-09-25 at 2 28 14 PM" src="https://github.com/user-attachments/assets/fb92c7dd-ce28-48fc-89e4-8f215ca09188">

`python -m hunting run-query --uuid 6e57e6a6-f150-405d-b8be-e4e666a3a86d`
<img width="904" alt="Screenshot 2024-09-25 at 2 31 58 PM" src="https://github.com/user-attachments/assets/dc19ac45-ef60-48d0-ad43-9dbf668cb573">


### View Hunt

` python -m hunting view-hunt --uuid 12526f14-5e35-4f5f-884c-96c6a353a544 --format json`
<img width="1608" alt="Screenshot 2024-09-25 at 2 29 43 PM" src="https://github.com/user-attachments/assets/b68f3c44-9db9-42a5-b1a9-1b53ff6a5b0b">

<img width="1331" alt="Screenshot 2024-09-25 at 2 30 10 PM" src="https://github.com/user-attachments/assets/35aaaa2e-19b0-4f23-81a1-5e1025c42e6f">

`python -m hunting view-hunt --uuid 12526f14-5e35-4f5f-884c-96c6a353a544 --query-only `
<img width="1331" alt="Screenshot 2024-09-25 at 2 30 10 PM" src="https://github.com/user-attachments/assets/cdb257b5-8f01-424f-a52b-54b9a4a94ffd">

</details>

<!--
  Summarize your PR. Animated gifs are 💯. Code snippets are ⚡️. Examples & screenshots are 🔥
-->

<details>
<summary>

## How To Test

</summary>

### Run Unit-Tests
`python -m unittest tests/test_hunt_data.py`

### Search for a specific tactic, technique or sub-technique ID in MITRE
`python -m hunting search --subtechnique T1078.004 --data-source aws`
`python -m hunting search --tactic TA0001 --data-source okta`
`python -m hunting search --technique T1078`

### Refresh the index (feel free to remove index entry and see if it populates again)
`python -m hunting refresh-index`

### Generate markdown
`python -m hunting generate-markdown /Users/tdejesus/code/src/detection-rules/hunting/okta/queries/initial_access_higher_than_average_failed_authentication.toml`
`python -m hunting generate-markdown okta`
`python -m hunting generate-markdown`

### Run Query
In order to do this, you must add a `.detection-rules-cfg.yaml` config with your Cloud ID and API key for authentication.

`python -m hunting run-query --file-path /Users/tdejesus/code/src/detection-rules/hunting/linux/queries/low_volume_external_network_connections_from_process.toml`
`python -m hunting run-query --uuid ef579900-75ef-11ef-b47f-f661ea17fbcc `
`python -m hunting run-query --uuid 6e57e6a6-f150-405d-b8be-e4e666a3a86d`


### View Hunt

`python -m hunting view-hunt --uuid 12526f14-5e35-4f5f-884c-96c6a353a544 --format json`
`python -m hunting view-hunt --uuid 12526f14-5e35-4f5f-884c-96c6a353a544 --query-only `

</details>

## Checklist

<!-- Delete any items that are not applicable to this PR. -->

- [x] Added a label for the type of pr: `bug`, `enhancement`, `schema`, `Rule: New`, `Rule: Deprecation`, `Rule: Tuning`, `Hunt: New`, or `Hunt: Tuning` so guidelines can be generated
- [ ] Added the `meta:rapid-merge` label if planning to merge within 24 hours
- [x] Secret and sensitive material has been managed correctly
- [x] Automated testing was updated or added to match the most common scenarios
- [x] Documentation and comments were added for features that require explanation

## Contributor checklist

- Have you signed the [contributor license agreement](https://www.elastic.co/contributor-agreement)?
- Have you followed the [contributor guidelines](https://github.com/elastic/detection-rules/blob/main/CONTRIBUTING.md)?

terrancedejesus commented on hunting/aws/docs/s3_public_bucket_rapid_object_access_attempts.md:
```diff
@@ -7,7 +7,7 @@
 - **Description:** This hunting query identifies when an anonymous user, outside of the known AWS IP ranges, makes multiple `GetObject` requests to a public S3 bucket. Rapid access to objects in a public S3 bucket may indicate an adversary attempting to exfiltrate data or perform reconnaissance on the bucket contents.
 
-- **UUID:** `ef244ca0-5e32-11ef-a8d3-f661ea17fbce`
+- **UUID:** `ef579900-75ef-11ef-b47f-f661ea17fbcc`
```
Duplicate UUID; unit test added

protectionsmachine commented:
<!-- add-pr-comment:guidelines-comment -->

## Enhancement - Guidelines

These guidelines serve as a reminder set of considerations when addressing adding a new schema feature to the code.

### Documentation and Context

- [ ] Describe the feature enhancement in detail (alternative solutions, description of the solution, etc.) if not already documented in an issue.
- [ ] Include additional context or screenshots.
- [ ] Ensure the enhancement includes necessary updates to the documentation and versioning.

### Code Standards and Practices

- [ ] Code follows established design patterns within the repo and avoids duplication.
- [ ] Code changes do not introduce new warnings or errors.
- [ ] Variables and functions are well-named and descriptive.
- [ ] Any unnecessary / commented-out code is removed.
- [ ] Ensure that the code is modular and reusable where applicable.
- [ ] Check for proper exception handling and messaging.

### Testing

- [ ] New unit tests have been added to cover the enhancement.
- [ ] Existing unit tests have been updated to reflect the changes.
- [ ] Provide evidence of testing and validating the enhancement (e.g., test logs, screenshots).
- [ ] Validate that any rules affected by the enhancement are correctly updated.
- [ ] Ensure that performance is not negatively impacted by the changes.
- [ ] Verify that any release artifacts are properly generated and tested.

### Additional Schema Related Checks

- [ ] Ensure that the enhancement does not break existing functionality. (e.g., run `make test-cli`)
- [ ] Review the enhancement with a peer or team member for additional insights.
- [ ] Verify that the enhancement works across all relevant environments (e.g., different OS versions).
- [ ] Confirm that all dependencies are up-to-date and compatible with the changes.
- [ ] Link to the relevant Kibana PR or issue provided
- [ ] Exported detection rule(s) from Kibana to showcase the feature(s)
- [ ] Converted the exported ndjson file(s) to toml in the detection-rules repo
- [ ] Re-exported the toml rule(s) to ndjson and re-imported into Kibana
- [ ] Updated necessary unit tests to accommodate the feature
- [ ] Applied min_compat restrictions to limit the feature to a specified minimum stack version
- [ ] Executed all unit tests locally with a test toml rule to confirm passing
- [ ] Included Kibana PR implementer as an optional reviewer for insights on the feature
- [ ] Implemented requisite downgrade functionality
- [ ] Cross-referenced the feature with product documentation for consistency
- [ ] Incorporated a comprehensive test rule in unit tests for full schema coverage
- [ ] Conducted system testing, including fleet, import, and create APIs (e.g., run `make test-remote-cli`)

terrancedejesus commented:
Putting this back into Draft, going to work on the following to finalize:

* Add a `run` command to fling queries at the API
* Add a disclaimer MD file

imays11 commented:
Great tool @terrancedejesus, awesome way to make the hunts more easily accessible for users! Something I noticed while using the hunting tool is that the file paths provided by the search command assume the working directory is the hunting directory. However, if you're using the tool from the root directory, `detection_rules`, then these links aren't as convenient to copy-paste > cat file. Maybe changing the links to be the full path starting at the root directory? This way even if someone is operating in the hunting folder, or another sub-folder, it's easier to copy the portion of the path they need, rather than someone having to `cd` to the hunting directory before conveniently using the file path.

terrancedejesus commented on Makefile:
```diff
@@ -26,6 +26,11 @@ deps: $(VENV)
+.PHONY: hunting
+hunting: $(VENV) deps
+	@echo "Installing all dependencies..."
+	$(PIP) install elasticsearch~=8.15.1 tabulate==0.9.0
```
This is a workaround for 8.12.1 being the version dependency by `detection_rules` and `lib/kibana`.

terrancedejesus commented:
@Mikaayenson @shashank-elastic @eric-forte-elastic - I have addressed the feedback recorded as noted below. Additionally, Mika had mentioned making the modules more modular and thus I have updated them to have an instantiation class with methods to handle most of the logic that is called from within each command. Thanks again for the feedback!

* Additional hunting tests
* Check ESQL client
* Check if get_hunt_path can be re-used elsewhere
* Do we remain consistency with MITRE or not (May not be required. The naming of the hunts are more important as the file name is never displayed.
* Bug in search with LLM (fixed)
* Add note in README about connecting to ES with config
* Address guidelines for hunting
* Textwrap tabulate output from running query (removed tabulate from `run-query` command)
* Change the numbering so that starts 1+
* Update README for Detection Rules README
* Check if we are introducing regressions
* Add test-hunting-cli
* Update makefile

terrancedejesus commented:
Nice addition of summary @Mikaayenson !

terrancedejesus commented on .github/PULL_REQUEST_GUIDELINES/hunt_new_guidelines.md:
```diff
@@ -5,27 +5,25 @@ Welcome to the `hunting` folder within the `detection-rules` repository! This di
-- [ ] `note` includes additional information (e.g., Triage and analysis investigation guides, timeline templates).
-- [ ] `tags` are relevant to the threat and align with `EXPECTED_HUNT_TAGS` in `definitions.py`.
-- [ ] `threat`, `techniques`, and `subtechniques` map to ATT&CK whenever possible.
+- [ ] `notes` includes additional information regarding data collected from the hunting query.
+- [ ] `mitre` matches appropriate technique and sub-technique IDs that hunting query collect's data for.
```
@Mikaayenson 

Description is there

`name` and `description` are descriptive and typo-free.

License is not, but we don't require it for rules either? Also if it's a community contribute, do they put Elastic license?

Mikaayenson replied:
It's a good question. I know its optional for our rules (but didn't see any without the Elastic license).

terrancedejesus replied:
Got it, if optional for rules then these would potentially inherit that.

terrancedejesus commented on hunting/README.md:
```diff
@@ -44,30 +44,82 @@ Otherwise, the names do not require the integration, since it is already annotat
+
+## Commands
+
+The `hunting` folder is a modularized library with it's own CLI via the user of [click](https://pypi.org/project/click/). All commands can be ran from the root of `detection-rules` repository as such: `python -m hunting COMMAND`.
```
Done! Also note, verbiage is tricky but I started referencing this as not just a folder but an executable package within detection rules based on everything we have done.

terrancedejesus commented on hunting/README.md:
```diff
@@ -44,30 +44,82 @@ Otherwise, the names do not require the integration, since it is already annotat
+   - Navigate to the respective folder (e.g., `aws/queries`, `macos/queries`) and create a new TOML file for your query.
+   - Ensure that the file is named descriptively, reflecting the purpose of the hunt (e.g., `credential_access_detection.toml`).
+
+2. **Add Relevant and Required Hunting Information**
```
I added a step to clone and pip install, but not sure if we need to recommend new make command. No preference.

Mikaayenson replied:
👍 No need. Just wondering how we let users know they can install this hunting logic using `make hunting-deps`.

terrancedejesus commented on hunting/__main__.py:
```diff
@@ -0,0 +1,212 @@
+
+
+@hunting.command('generate-markdown')
+@click.argument('path', required=False)
+def generate_markdown(path):
```
Added `click.Path`

terrancedejesus commented on hunting/__main__.py:
```diff
@@ -0,0 +1,212 @@
+    click.echo("Searching for queries based on provided filters...")
+
+    # Filter out None values from the MITRE filter tuple
+    mitre_filters = tuple(filter(None, (tactic, technique, sub_technique)))
```
Good question. MITRE is something consistent to filter on because it's required but also adds a constraint to what can be searched based on a popular framework.

We can add text-based searching from the notes, also tags is another option if we implemented them. I assume people may want to search for anything `powershell` or `cloud` or `Lumma` too. That is a good point but will likely require us introduce tags in a later PR (since it will require us to update ALL hunting queries. For now, we can do a `keyword` option that searches in description, notes and name.

terrancedejesus commented on hunting/__main__.py:
```diff
@@ -0,0 +1,212 @@
+    if output_format == 'toml':
+        click.echo(hunt_path.read_text())
+    elif output_format == 'json':
+        import json
```
derp, thanks!

terrancedejesus commented on hunting/__main__.py:
```diff
@@ -0,0 +1,276 @@
+
+        # Update language-based summary
+        for language in languages:
+            if language == 'SQL':
```
SQL is the query language used for OSQuery. To me it would be like saying "ES|QL, Discover" where discover is the tool but KQL is the actual language used to query the data behind the scenes.

Maybe we can add a note that if SQL, then it is meant for OSQuery. I think SQL is also universal to some degree and can be used for big data searches in Snowflake, BigQuery, etc. too depending on ingest of course.

Mikaayenson replied:
👍 This line just renders OSQuery in the stats summary table vs SQL.

terrancedejesus replied:
Updated the README so that we note within language explanation that if SQL is observed in TOML files or markdown, then this relates to OSQuery.

terrancedejesus commented on hunting/utils.py:
```diff
@@ -0,0 +1,134 @@
+    save_index_file(base_path, directories)
+
+
+def filter_elasticsearch_params(config: dict) -> dict:
```
We cast anything from the config file to `detection_rules.misc.get_elasticsearch_client`. In doing this, I accidentally left the `KIBANA_URL` in there from a previous config. As such it threw an error as that was for the kibana client. Therefore, rather than require users to adjust existing configs, I settled with filtering with what is defined as acceptable parameters of `get_elasticsearch_client`. Happy to adjust if need be.

Mikaayenson replied:
make sense.

terrancedejesus commented on pyproject.toml:
```diff
@@ -52,7 +53,7 @@ dev = ["pep8-naming==0.13.0", "PyGithub==2.2.0", "flake8==7.0.0", "pyflakes==3.2
 [tool.setuptools]
 package-data = {"kql" = ["*.g"]}
-packages = ["detection_rules", "rta"]
+packages = ["detection_rules", "rta", "hunting"]
```
I added it as a package based on CLI functionality. Now that a user can search for the write query and run against existing stack, I imagined there may be a user that simply wants to use this executable package manually or in CI/CD, etc. and does not use the detection rules. It also mimics RTA in that it is it's own executable package.

terrancedejesus commented on tests/test_hunt_data.py:
```diff
@@ -63,6 +62,59 @@ def test_markdown_existence(self):
+        cls.hunting_index = load_index_file()
+
+    def test_mitre_techniques_present(self):
+        """Ensure each query has at least one MITRE technique."""
```
Good catch. The search functionality relies on MITRE, so it's almost required. But in practicality, not everything may be mapped to MITRE. 

We can make mitre optional and keep the unit test or we can keep it optional and remove. I'm fine with either but am leaning towards making it a requirement.

terrancedejesus replied:
<img width="2271" alt="Screenshot 2024-09-30 at 2 50 53 PM" src="https://github.com/user-attachments/assets/9ba89786-44db-40c6-861a-f67feb132a04">

Mikaayenson commented on hunting/README.md:
```diff
@@ -44,30 +44,90 @@ Otherwise, the names do not require the integration, since it is already annotat
+
+## Commands
+
+The `hunting` folder is an executable package with it's own CLI using [click](https://pypi.org/project/click/). All commands can be ran from the root of `detection-rules` repository as such: `python -m hunting COMMAND`.
```
```suggestion
The `hunting` folder is an executable package with its own CLI using [click](https://pypi.org/project/click/). All commands can be ran from the root of `detection-rules` repository as such: `python -m hunting COMMAND`.
```

eric-forte-elastic commented on detection_rules/etc/test_hunting_cli.bash:
```diff
@@ -0,0 +1,34 @@
+python -m hunting hunt-summary --breakdown platform
+
+echo "Generating summary of hunts by language"
+python -m hunting hunt-summary --breakdown language
```
Nit, missing newline

eric-forte-elastic commented on detection_rules/etc/test_hunting_cli.bash:
```diff
@@ -0,0 +1,34 @@
+python -m hunting hunt-summary --breakdown platform
+
+echo "Generating summary of hunts by language"
+python -m hunting hunt-summary --breakdown language
```
```suggestion
python -m hunting hunt-summary --breakdown language

```

eric-forte-elastic replied:
Nit, missing newline

eric-forte-elastic commented on hunting/search.py:
```diff
@@ -0,0 +1,163 @@
+        self.mitre_technique_ids = set()
+        self.reverse_tactics_map = {v: k for k, v in tactics_map.items()}
+
+    def process_mitre_filter(self, mitre_filter: tuple):
```
Should this method be private as well? Fine either way, just seeing that it is only called internally and does not return anything.

terrancedejesus replied:
Good catch, updating!

eric-forte-elastic reviewed (approved):
Great work with this @terrancedejesus ! :rocket: Great peer review. LGTM :+1:

Mikaayenson reviewed (approved):
🚢

traut commented on hunting/__main__.py:
```diff
@@ -0,0 +1,278 @@
+            click.echo(f"Generating Markdown for folder: {path}")
+            markdown_generator.process_folder(path)
+        else:
+            click.echo(f"Invalid path: {path}. It should be a valid TOML file or a folder.")
```
It would be good to `raise ValueError("....")` here to stop the execution flow -- the path was provided but it is not valid

terrancedejesus replied:
Fixed!

traut commented on hunting/__main__.py:
```diff
@@ -0,0 +1,278 @@
+
+@hunting.command('generate-markdown')
+@click.argument('path', required=False)
+def generate_markdown(path: click.Path):
```
type hint nitpick: is `path` is optional, the type should be `path: click.Path | None`. This would allow pyright to check the risky access to the var in the function.

on another note -- I found it a bit easier to convert paths directly into `pathlib.Path`, with this config for `click` -- https://github.com/elastic/react/blob/b22db38b245286d3842c074c7088257c26685ef7/react/main.py#L61-L66

terrancedejesus replied:
Addressed!

traut commented on hunting/__main__.py:
```diff
@@ -0,0 +1,278 @@
+    hunt_path, error_message = get_hunt_path(uuid, path)
+
+    if error_message:
+        click.secho(f"{error_message}", fg="red", bold=True)
```
wow, TIL that `click` can do formatting in terminal!

traut commented on hunting/__main__.py:
```diff
@@ -0,0 +1,278 @@
+    mitre_filters = tuple(filter(None, (tactic, technique, sub_technique)))
+
+    # Call the search method of QueryIndex with the provided MITRE filters, data_source, and keyword
+    results = query_index.search(mitre_filter=mitre_filters, data_source=data_source, keyword=keyword)
```
btw, thank you very much for all the inline comments in the code, it is greatly appreciated!

traut commented on hunting/__main__.py:
```diff
@@ -0,0 +1,278 @@
+                language = 'OSQuery'
+            if language not in language_summary:
+                language_summary[language] = 0
+            language_summary[language] += 1
```
The counters here might be easier to implement with `collections.Counter()`. Take a look at the example [here](https://github.com/elastic/cortado/blob/2c973f564c0bb1e1fd7e0ecd5a9d5d9f6d03a191/cortado/cli.py#L107-L111)

terrancedejesus replied:
Thanks for the example, implemented!

traut commented on hunting/markdown.py:
```diff
@@ -0,0 +1,144 @@
+            markdown += "\n\n## References\n\n" + "\n".join(f"- {ref}" for ref in hunt_config.references)
+
+        markdown += f"\n\n## License\n\n- `{hunt_config.license}`\n"
+        return markdown
```
for the follow-up work -- it would be great to have a Jinja2 template for this!

terrancedejesus replied:
That's actually a great idea! Would love to maybe make it another feature update at a later time though. Can create an issue with that for future reference.

traut reviewed (approved):
that's a lot of work, very cool! I left some comments - nothing major, so up to you if you would like to address them. 

2 general observations:
- I think we should stop using `click` as a logging solution (we do the same in REACT). It's great for parsing CLI arguments (though I'm using `typer` in [cortado](https://github.com/elastic/cortado) for simplicity), but it lacks the flexibility and power a proper logging library would provide -- my favourite here is `structlog`. 
- if we're setting type hints we might as well enforce the correctness with `pyright`. I see it in REACT as well. Without enforcement, they are just there for decoration :)

terrancedejesus commented:
@traut 

> I think we should stop using click as a logging solution (we do the same in REACT). It's great for parsing CLI arguments (though I'm using typer in [cortado](https://github.com/elastic/cortado) for simplicity), but it lacks the flexibility and power a proper logging library would provide -- my favourite here is structlog.

We have discussed logging in detection rules for awhile now and I agree, proper logging would be great for using the repo or the many supported features such as RTA, Detection Rules, Hunting, etc. This would be a good future improvement, but IMO not to be introduced in this PR. 

> if we're setting type hints we might as well enforce the correctness with pyright. I see it in REACT as well. Without enforcement, they are just there for decoration :)

Agreed for enforcement, but out-of-scope for this PR if I understand. More as an overall repository addition since it requires introducing pyright to the repo, adding a config, adjusting CI/CD builds, etc.
